### DISCUSSION TITLE
Provide assets in a lightweight way?

### DISCUSSION THREAD

**Discussion by jaynetics** [activeadmin/activeadmin#8246] [General]
i think it could benefit activeadmin's popularity to have less dependencies.

with more and more rails apps being API-only, activeadmin is now probably one of the major reasons why people set up sprockets in their apps, which seems a bit wasteful just to compile JS that doesn't usually change.

on the CSS side, installing sassc always used to be a bit of a nuisance with its tricky compilation. the activeadmin v4 way, however, of having to install node + node modules + running a yarn watcher alongside rails, feels like an even bigger burden, especially if this is just to compile some CSS that doesn't usually change, and it does lead to some struggles as seen in #7947, #8223, #8226.

people have come up with hacks to work around this. [this blog](https://henrikbjorn.medium.com/til-how-to-use-activeadmin-v4-without-installing-nodejs-ccdaf3ace005) describes how to use tailwindcss-rails to circumvent the node requirement, restoring roughly the level of "comfort" that sassc used to provide. some people have [forked](https://github.com/j-manu/active_admin) AA with the assets included. others have monkeypatched the asset tag helpers to render CDN URLs instead. what would be ideal, IMHO, was if AA itself were to offer some lightweight and convenient way to render itself ...

**Comment by jaynetics** [comment]
struggles with asset provisioning keep cropping up: #8302, #8361.

i've also come to realize another advantage of an assets-included mode: it would be possible to have single-file setups for demonstrations and bug reproduction. AFAICT the [current template](https://github.com/activeadmin/activeadmin/blob/master/tasks/bug_report_template.rb) is missing assets.

**Reply by javierjulio** [reply]
This contradicts your other statement though where you'd want fewer dependencies. The problem with this approach is we have to lock in at least two dependencies for CSS and JS whether you use it or not. There are many options and people don't want to include what they don't use. I think going with one approach would be easier and ideal.

I'm not sure the bug report template script would be ideal for that. Rails doesn't use assets in there scripts. It was easier before because of Sprockets and asset gems. A lot has changed.

**Reply by jaynetics** [reply]
@javierjulio i don't quite understand how this would add dependencies, but perhaps my idea on how to include the assets wasn't clear.

it is roughly this:

- hook into the gem build step to build the assets and include the resulting files in the gem
- add a configuration option to use these pre-built assets, e.g. `use_precompiled_assets`
- add a rack middleware to provide the pre-built assets to the frontend, [like in sidekiq](https://github.com/sidekiq/sidekiq/blob/a9cab7d047230bdc53c35c9ead849783c758a3b1/lib/sidekiq/web.rb#L167-L181) and other engines
- in the html_head, if `use_precompiled_assets` is true, link to the pre-built assets, else do the current thing

this would require zero asset-related dependencies on the end-user side. i'd be willing to try to do this in a PR, or maybe make my own gem that does this if you consider it scope creep.

**Reply by javierjulio** [reply]
The problem with this approach is it would conflict with any user style changes on the app side. They wouldn't be able to make Tailwind changes and have it apply safely either, it would conflict. I made a [deliberate choice not to do this](https://github.com/activeadmin/activeadmin/discussions/8223#discussioncomment-9036575). It makes sense for Sidekiq because the styles are not meant to change but thats not the case here. You are more than welcome to try something in your fork and share a POC. I'd love to see it in action. Thanks.

**Comment by jaynetics** [comment]
I've made a gem to do this: [activeadmin_assets](https://github.com/jaynetics/activeadmin_assets/).
